The integration of SasView

AtomAnu opened the pull request:
#### Description of Work
This is the integration of SasView. There are five main additions for this integration.

1. An additional example script, `example_runScripts_SasView.py` has been added to use               SasView as the software to fit the user-selected problem(s). At the start of the example script,   there are three condition statements with each statement verifying if one of the three required SasView packages is installed. Bumps, sasmodels and sascalc are the three required packages. As explained in the FitBenchmarking README.md file, Bumps and sasmodels can be installed using `pip` while sascalc is not an independent package.

   Additionally, the package sasmodels contains a script `data.py` which contains a module   called `load_data`. This module requires sascalc in order to function. In this module's code, the line of code which imports sascalc is written below.

   `from sas.sascalc.dataloader.loader import Loader`

   Furthermore, sascalc now exists under the directory `src/sas` which resides in the SasView repository (https://github.com/SasView/sasview)

   As a consequence, it was neccessary to copy the folder `sas` from the SasView repository and place it inside FitBenchmarking (`fitbenchmarking/sas`) in order to run SasView headless inside FitBenchmarking.

2. A parsing script, `parse_sasview_data.py` has been added to parse SasView problem definition files and their corresponding data files. This script can later be integrated with `parse_fitbenchmark_data.py`.

   in `parse_sasview_data.py`, the eval_f and get_function modules work in the same manner as that of `parse_nist_data.py` and `parse_fitbenchmark_data.py`. However, the module eval_f for `parse_sasview_data.py` accepts the parameters list in two formats. The first format is a string which contains the parameter names and values. The second format is to accept the parameter values as multiple parameters to the module.

   In `parse_fitbenchmark_data.py`, a new module, `get_bumps_function`, is created to format Mantid functions into the functions that are acceptable by Bumps. Functions/models that are acceptable by Bumps should not contain any arguments (*arg) or keyword arguments (**kwargs).

3. A data preparation script, `sasview/prepare_data.py`, has been created to prepare the data in the correct format. There are three main tasks done in this script. The first task is to set the data errors (data_e) if they are not given in the data file. The second task is to set the X data range if specified. If the X data range is not specified, the range would be set from the minimum X value to the maximum X value. The final task is to create a data object of type `sasmodels.data.Data1D` so that it can be passed onto Bumps.

4. A function definition preparation script, `sasview/func_def.py`, has been created to prepare the models/functions and their starting parameter values for Bumps fitting. The script also prepares the initial and the final function definition strings for each fit.

5. Finally, a script, `sasview/main.py`, has been added to perform fitting using Bumps. All the fitting occur in the module `fit`. One can observe that this module is quite lengthy. This is because the method of configuring the function/model is different for SasView models and other functions from Scipy and Mantid. Nevertheless, this module could be shorten up in the future.

   This script also calculates the chi-squared value and the run time of each fit. Then, a result object is created for the graph and table creating modules. A single result object contains the fitted Y values, the fit status, the chi-squared value, the fit runtime, the minimiser name, the initial function definition string and the final function definition string.

***
Note that the unit tests for these scripts have not been completed. 
#### Testing Instructions

1. run `example_runScripts.py` with problems of type NIST, FitBenchmark and SasView
2.  run `example_runScripts_mantid.py` with problems of type NIST, FitBenchmark and SasView
3. run `example_runScripts_SasView.py` with problems of type NIST, FitBenchmark and SasView

Anders-Markvardsen commented:
@AtomAnu thanks for this PR and it detailed description. You write "As a consequence, it was neccessary to copy the folder sas from the SasView repository and place it inside FitBenchmarking (fitbenchmarking/sas) in order to run SasView headless inside FitBenchmarking.". 

A questions:

* In the sasview meeting tuesday did you discuss with the sasview team this aspects? - i.e. to include a copy of parts of there code until the sasview team has separated sascal out as separate package.
  * assuming this is OK then as a minimum please add readme in that folder to specify when the copy was made and a copy of your above reasoning for adding it.

Obviously the above is not ideal, but with the sasview team actively working on separating out sascal it seems like the best workaround

Anders-Markvardsen commented:
Thinking about it again. As an alternative to copy and paste have you considered code which

* clone the sasview repo in read only mode, either a particular version of it directly or where after the repo has been cloned it is reset to a particular version. The latter is to guard against the expected future changes of this part of the sasview codebase 
* copy the sascal part, and this might be a tricky part: to a location where it is visible from the python path as appropriate 
* delete the cloned sasview repo

See https://stackoverflow.com/questions/2472552/python-way-to-clone-a-git-repository for suggestions for how a github repo can be cloned

Anders-Markvardsen commented on README.md:
```diff
@@ -28,10 +28,10 @@ For help of how to use the command line/terminal, click the hyperlink correspond
 
 ## FitBenchmarking Scipy
-The `example_runScripts.py` file can be changed such that it benchmarks minimizers supported by scipy instead of mantid (details provided in the file itself).
+The `example_runScripts.py` file is designed to benchmark minimizers supported by scipy instead of mantid (details provided in the file itself).
```
```suggestion
The `example_runScripts.py` file is designed to benchmark minimizers supported by software/libraries that provide straightforward cross-platform Python install; as of know this mean SciPy (more details provided in the file itself).
```

Anders-Markvardsen commented on example_scripts/example_runScripts.py:
```diff
@@ -77,7 +78,7 @@
 # problem_sets = ["CUTEst", "Muon_data", "Neutron_data", "NIST/average_difficulty", "NIST/high_difficulty", "NIST/low_difficulty"]
 
-problem_sets = ["NIST/low_difficulty"]
+problem_sets = ["Muon_data"]
```
@AtomAnu did you mean to change problem set up Muon_data in this script? I.e. with problem_sets = Muon_data this script with not run without mantid

AtomAnu commented:
Dear Anders,

The intention was for the purpose of testing. The problem_sets variable
should be set to [“NIST/low_difficulty”].


Kind regards,
Atom

On Fri, Aug 23, 2019 at 8:08 AM Anders Markvardsen <notifications@github.com>
wrote:

> *@Anders-Markvardsen* commented on this pull request.
> ------------------------------
>
> In example_scripts/example_runScripts.py
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199#discussion_r316999713>
> :
>
> > @@ -77,7 +78,7 @@
>  # problem_sets = ["Neutron_data", "NIST/average_difficulty"]
>  # problem_sets = ["CUTEst", "Muon_data", "Neutron_data", "NIST/average_difficulty", "NIST/high_difficulty", "NIST/low_difficulty"]
>
> -problem_sets = ["NIST/low_difficulty"]
> +problem_sets = ["Muon_data"]
>
> @AtomAnu <https://github.com/AtomAnu> did you mean to change problem set
> up Muon_data in this script? I.e. with problem_sets = Muon_data this script
> with not run without mantid
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199?email_source=notifications&email_token=AMN6IMHOT4NINSWMBS57NALQF6EHPA5CNFSM4IOIDXA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCPJMTA#pullrequestreview-278828620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMN6IMH4FIDICPCPYETLQC3QF6EHPANCNFSM4IOIDXAQ>
> .
>

Anders-Markvardsen commented:
Note I observe
```
C:\Midlertiddig\fitbenchmarking\example_scripts>python example_runScripts_SasView.py

Running the benchmarking on the 1D_data problem set

Traceback (most recent call last):
  File "example_runScripts_SasView.py", line 117, in <module>
    use_errors=use_errors, results_dir=results_dir)
  File "C:\Midlertiddig\fitbenchmarking\fitbenchmarking\fitting_benchmarking.py", line 65, in do_fitting_benchmark
    prob_results = do_benchmarking(user_input, problem_groups, group_name)
  File "C:\Midlertiddig\fitbenchmarking\fitbenchmarking\fitting_benchmarking.py", line 85, in do_benchmarking
    prob_results.append(do_fitbm_group(user_input, block))
  File "C:\Midlertiddig\fitbenchmarking\fitbenchmarking\fitting_benchmarking.py", line 115, in do_fitbm_group
    problem = parse.parse_problem_file(prob_file)
  File "C:\Midlertiddig\fitbenchmarking\fitbenchmarking\parsing\parse.py", line 59, in parse_problem_file
    problem = parse_sasview_data.FittingProblem(prob_file)
  File "C:\Midlertiddig\fitbenchmarking\fitbenchmarking\parsing\parse_sasview_data.py", line 33, in __init__
    data_obj = load_data(data_file_path)
  File "C:\MantidInstall\bin\lib\site-packages\sasmodels\data.py", line 54, in load_data
    from sas.sascalc.dataloader.loader import Loader  # type: ignore
  File "C:\Midlertiddig\fitbenchmarking\sas\sascalc\dataloader\loader.py", line 36, in <module>
    from .readers import cansas_reader
  File "C:\Midlertiddig\fitbenchmarking\sas\sascalc\dataloader\readers\cansas_reader.py", line 13, in <module>
    from lxml import etree
ImportError: No module named lxml
```

AtomAnu commented:
@Anders-Markvardsen I forgot to include that you would be required to install the package **lxml** (`pip install lxml`) as it is required by sascalc in the case when it is reading .xml files.

Anders-Markvardsen commented:
Concerning the new folder in benchmark_problems '1D_data' I would suggest changing the structure to

SAS_modelling/1D/data_files/
                               prob_def_1.txt
                               prob_def_2.txt

SAS stands for Small Angle Scattering and the examples/problems in that folder are specifically for modelling SAS data

AtomAnu commented:
@Anders-Markvardsen 
I have just added the following changes:
1. Updated the FitBenchmarking README.md file in the section, **FitBenchmarking SasVIew**, to include the instruction to install **lxml**.
2. Added a README.md file inside the folder **sas** to explain why `sascalc` is required in FItBenchmarking and why it is added here temporarily.
3. Edited the format of log messages sent when fits fail. The level of the log messages has also been changed to `warning` instead of `error`.
4. Changed the name of folder containing 1D SasView problem files from `1D_data` to `1D`. The location of the folder is also changed. It is now located in a folder called `SAS_modelling`.
5. Changed the folder names, `Muon_data` and `Neutron_data`, to `Muon` and `Neutron`, respectively.

Anders-Markvardsen commented on example_scripts/SasView_example.py:
```diff
@@ -0,0 +1,87 @@
+from sasmodels.core import load_model
```
What is the purpose of this file and how does it differ from example_runScripts_SasView.py

Anders-Markvardsen commented on fitbenchmarking/fitbenchmark_one_problem.py:
```diff
@@ -58,8 +58,11 @@ def fitbm_one_prob(user_input, problem):
         # scipy does not currently support the GEM problem
-        if 'GEM' in problem.name and user_input.software == 'scipy':
-            break
+        # if 'GEM' in problem.name and user_input.software == 'scipy':
```
please remove dead code comments

AtomAnu commented:
That script is a script in which I have added in order to test running
SasView headless from the start to the end. I have forgotten to remove it
when creating the pull request.

On Fri, Aug 23, 2019 at 7:51 PM Anders Markvardsen <notifications@github.com>
wrote:

> *@Anders-Markvardsen* commented on this pull request.
> ------------------------------
>
> In example_scripts/SasView_example.py
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199#discussion_r317261155>
> :
>
> > @@ -0,0 +1,87 @@
> +from sasmodels.core import load_model
>
> What is the purpose of this file and how does it differ from
> example_runScripts_SasView.py
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199?email_source=notifications&email_token=AMN6IMGDPVD3Y4NNBG5DNV3QGAWTPA5CNFSM4IOIDXA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCR4H6I#pullrequestreview-279167993>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMN6IMH6ADA4GBXDZY5VX7DQGAWTPANCNFSM4IOIDXAQ>
> .
>

Anders-Markvardsen commented on fitbenchmarking/fitting/mantid/externals.py:
```diff
@@ -61,9 +61,15 @@ def set_ties(function_object, ties):
     for idx, ties_per_func in enumerate(ties):
         for tie in ties_per_func:
-            exec "function_object.tie({'f" + str(idx) + "." + tie + "})"
+            exec("param_dict = {'f" + str(idx) + "." + tie + "}")
```
param_dict appears to be declared but not used. Please delete commented out lines of code that are no longer needed. Please add a code comment to explain this more complex code

Anders-Markvardsen commented on fitbenchmarking/fitting/mantid/func_def.py:
```diff
@@ -100,6 +102,47 @@ def function1D(self, xdata):
+    Helper function that parses the SasView function definitions and
+    transforms them into a mantid-readable format.
+
+    @param problem :: object holding the problem information
```
change to 'object holding the problem information in SasView format' ?

Anders-Markvardsen commented on fitbenchmarking/fitting/mantid/func_def.py:
```diff
@@ -100,6 +102,47 @@ def function1D(self, xdata):
+
+    @param problem :: object holding the problem information
+
+    @returns :: the formatted function definition (str)
```
change to 'Mantid formatted function definition (str)'  ?

Anders-Markvardsen commented on fitbenchmarking/fitting/mantid/main.py:
```diff
@@ -152,7 +152,7 @@ def parse_result(fit_result, t_start, t_end):
         fin_function_def = str(fit_result.Function)
         runtime = t_end - t_start
-
+    print(fin_function_def)
```
Do you mean to have this print statement here?

Anders-Markvardsen commented on fitbenchmarking/fitting/plotting/plots.py:
```diff
@@ -230,3 +233,29 @@ def get_mantid_starting_guess_data(wks_created, function, problem):
+    :return:
+    """
+
+    # yData = function[0](data_struct.x)
```
dead comment?

Anders-Markvardsen commented on fitbenchmarking/fitting/plotting/plots.py:
```diff
@@ -230,3 +233,29 @@ def get_mantid_starting_guess_data(wks_created, function, problem):
+    # yData = function[0](data_struct.x)
+
+    yData = problem.eval_f(data_struct.x, function[1])
+    # from sasmodels.bumps_model import Experiment, Model
```
dead comment lines?

AtomAnu commented:
That module is currently in progress of investigation regarding the issue
of setting ties on mantid. It was initially used for the module ties for
mantid. However, the modules ties does not seem to set the specified ties.
Thus, I have tried to use the module fix instead as all the current ties
specified in the problem definition files are for the purpose of fixing
some parameter values.

I have paused the investigation at the moment as the module fix does not
work as well when fitting mantid problems with Scipy and SasView.

Could you give me an advice of what I should do? Should I remove any
statements that call this module?

On Fri, Aug 23, 2019 at 7:59 PM Anders Markvardsen <notifications@github.com>
wrote:

> *@Anders-Markvardsen* commented on this pull request.
> ------------------------------
>
> In fitbenchmarking/fitting/mantid/externals.py
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199#discussion_r317263722>
> :
>
> >      for idx, ties_per_func in enumerate(ties):
>          for tie in ties_per_func:
> -            exec "function_object.tie({'f" + str(idx) + "." + tie + "})"
> +            exec("param_dict = {'f" + str(idx) + "." + tie + "}")
>
> param_dict appears to be declared but not used. Please delete commented
> out lines of code that are no longer needed. Please add a code comment to
> explain this more complex code
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199?email_source=notifications&email_token=AMN6IMCWL7GS4XA52J3KSODQGAXQFA5CNFSM4IOIDXA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCR5DJQ#pullrequestreview-279171494>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMN6IMDDFTXWFX5PJTZ5H2LQGAXQFANCNFSM4IOIDXAQ>
> .
>

Anders-Markvardsen commented on fitbenchmarking/fitting/plotting/plots.py:
```diff
@@ -230,3 +233,29 @@ def get_mantid_starting_guess_data(wks_created, function, problem):
+def get_sasview_starting_guess_data(data_struct, problem, function):
+    """
+
+    :param data_struct:
```
please add some minimum info about input parameters and the return output

Anders-Markvardsen commented on fitbenchmarking/fitting/prerequisites.py:
```diff
@@ -45,6 +45,8 @@ def prepare_software_prerequisites(software, problem, use_errors):
         return prepare_scipy(problem, use_errors)
+    elif software == 'sasview':
+        return prepare_sasview(problem, use_errors)
     # elif software == 'your_software':
```
dead comments?

Anders-Markvardsen commented on fitbenchmarking/fitting/prerequisites.py:
```diff
@@ -95,3 +97,14 @@ def prepare_scipy(problem, use_errors):
         problem.end_x = np.inf
     return data, cost_function, function_definitions
+
+def prepare_sasview(problem, use_errors):
```
prepare sasview for what? Please explain in comment

AtomAnu commented:
Sure! I’ll change the parts which you have mentioned accordingly.

On Fri, Aug 23, 2019 at 8:09 PM Anders Markvardsen <notifications@github.com>
wrote:

> *@Anders-Markvardsen* commented on this pull request.
> ------------------------------
>
> In fitbenchmarking/fitting/prerequisites.py
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199#discussion_r317267145>
> :
>
> > @@ -95,3 +97,14 @@ def prepare_scipy(problem, use_errors):
>          problem.start_x = - np.inf
>          problem.end_x = np.inf
>      return data, cost_function, function_definitions
> +
> +def prepare_sasview(problem, use_errors):
>
> prepare sasview for what? Please explain in comment
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fitbenchmarking/fitbenchmarking/pull/199?email_source=notifications&email_token=AMN6IMGTGRMVMCBVKW27PMTQGAYYBA5CNFSM4IOIDXA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCR6HJY#pullrequestreview-279176103>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMN6IMG72XHGEBTMCB5MOOTQGAYYBANCNFSM4IOIDXAQ>
> .
>

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/func_def.py:
```diff
@@ -0,0 +1,140 @@
+
+def function_definitions(problem):
+    """
+    Transforms the prob.equation field into a function that can be
```
Is this def description correct - i.e. does it match what this def does?

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/func_def.py:
```diff
@@ -0,0 +1,140 @@
+
+    return problem_type
+
+def get_fin_function_def(final_param_values, problem, init_func_def):
```
what does this def do? and what does 'fin' in get_fin_function_def stand for?

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/func_def.py:
```diff
@@ -0,0 +1,140 @@
+
+def get_init_function_def(function, problem):
+    """
+    Get the initial function definition string.
```
Do you mean 'Create and return the initial function definition string suitable to be passed to bumps.' ?

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/func_def.py:
```diff
@@ -0,0 +1,140 @@
+
+    problem_type = extract_problem_type(problem)
+
+    if not 'name=' in str(problem.equation):
```
Please add code comment to explain which case 'if not 'name=' in str(problem.equation):' refers to

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/func_def.py:
```diff
@@ -0,0 +1,140 @@
+    elif problem_type == 'SasView'.upper():
+        init_function_def = problem.equation + ',' + problem.starting_values
+        init_function_def = re.sub(r"(=)([-+]?\d+)([^.\d])", r"\g<1>\g<2>.0\g<3>", init_function_def)
+    else:
```
And which case this refers to

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/func_def.py:
```diff
@@ -0,0 +1,140 @@
+        init_function_def = function[2] + " | " + param_string
+    elif problem_type == 'SasView'.upper():
+        init_function_def = problem.equation + ',' + problem.starting_values
+        init_function_def = re.sub(r"(=)([-+]?\d+)([^.\d])", r"\g<1>\g<2>.0\g<3>", init_function_def)
```
this looks tricky - a code comment would help

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/main.py:
```diff
@@ -0,0 +1,171 @@
+    return results_problem, best_fit
+
+def fit(problem, data, function, minimizer, init_func_def):
+    """
```
code comment please

Anders-Markvardsen replied:
I meant def comment please

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/main.py:
```diff
@@ -0,0 +1,171 @@
+    t_start, t_end = None, None
+    model = function[0]
+
+    if hasattr(model, '__call__'):
```
A few code comments here may help

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/main.py:
```diff
@@ -0,0 +1,171 @@
+
+    final_param_values = result.x
+
+    fin_func_def = get_fin_function_def(final_param_values, problem, init_func_def)
```
what does 'fin' in fin_func_def stand for

AtomAnu replied:
The word fin stands for final in this case.

Anders-Markvardsen commented on fitbenchmarking/fitting/sasview/prepare_data.py:
```diff
@@ -0,0 +1,91 @@
+
+from utils.logging_setup import logger
+
+def prepare_data(problem, use_errors):
```
prepare data for what?

Anders-Markvardsen commented on fitbenchmarking/fitting/scipy/func_def.py:
```diff
@@ -51,37 +51,50 @@ def get_fin_function_def(init_function_def, func_callable, popt):
         popt = list(popt)
         params = init_function_def.split("|")[1]
-        params = re.sub(r"[-+]?\d+\.\d+", lambda m, rep=iter(popt):
+        params = re.sub(r"[-+]?\d+[.]\d+", lambda m, rep=iter(popt):
```
looks subtle - code comment please and e.g. line 61 also

Anders-Markvardsen commented on fitbenchmarking/fitting/scipy/func_def.py:
```diff
@@ -51,37 +51,50 @@ def get_fin_function_def(init_function_def, func_callable, popt):
+
+    problem_type = extract_problem_type(problem)
+
+    if not 'name=' in str(problem.equation):
```
please add comment to specify what use case this is. And for else below also

Anders-Markvardsen commented on fitbenchmarking/parsing/fitbenchmark_data_functions.py:
```diff
@@ -50,11 +50,36 @@ def fitbenchmark_func_definitions(functions_string):
     return function_defs
 
+def get_fit_function_without_kwargs(fit_function, functions_string):
+    """
```
please add short def description

Anders-Markvardsen commented on fitbenchmarking/parsing/fitbenchmark_data_functions.py:
```diff
@@ -50,11 +50,36 @@ def fitbenchmark_func_definitions(functions_string):
+def get_fit_function_without_kwargs(fit_function, functions_string):
+    """
+
+    :param fit_function:
```
some parameter info please

Anders-Markvardsen commented on fitbenchmarking/parsing/fitbenchmark_data_functions.py:
```diff
@@ -179,6 +204,7 @@ def get_fitbenchmark_ties(param_set, ties):
         else:
             tie = param_set[start + 1:comma]
         ties_per_function.append(tie.replace("=", "': "))
+        # ties_per_function.append(tie)
```
dead comment?

Anders-Markvardsen commented on fitbenchmarking/parsing/parse_fitbenchmark_data.py:
```diff
@@ -59,7 +59,7 @@ def __init__(self, fname):
         self._equation = entries['function']
 
-        self._starting_values = None
+        # self._starting_values = (entries['function'].split(',', 1))[1]
```
dead comment?

Anders-Markvardsen commented:
All definitions works with Scipy and Mantid. All work for SasView except for FitBenchmark type problems. For example by setting problem_sets = ["Neutron"] in example_runscript_sasview gives the error

```
C:\Midlertiddig\fitbenchmarking\fitbenchmarking\fitting\sasview\main.py in fit(problem, data, function, minimizer, init_func_def)
    107             # The problem type is NIST
    108             #Formatting the function parameter names
--> 109             formatted_param_names = [param[0] for param in problem.starting_values]
    110
    111         param_values = function[1]

TypeError: 'NoneType' object is not iterable
```

Anders-Markvardsen commented:
@AtomAnu thanks for the changes and again for your work over the summer.

I have created work for follow up work that this PR has highlighted during testing.
